pandas/_libs/lib.pyx

```diff
@@ -1698,6 +1698,48 @@ cpdef bint is_string_array(ndarray values, bint skipna=False):
     return validator.validate(values)
 
 
+cpdef ndarray ensure_string_array(
+        values, object na_value=np.nan, bint convert_na_value=True, bint copy=True):
+    """Returns a new numpy array with object dtype and only strings and na values.
+
+    Parameters
+    ----------
+    values : array-like
+        The values to be converted to str, if needed
+    na_value : Any
+        The value to use for na. For example, np.nan or pd.NAN
+    convert_na_value : bool, default True
+        If False, existing na values will be used unchanged in the new array
+    copy : bool, default True
+        Whether to wnsure that a new array is returned
+
+    Returns
+    -------
+    ndarray
+    """
+    cdef:
+        Py_ssize_t i = 0, n = len(values)
+
+    result = np.asarray(values, dtype="object")
+    if copy and result is values:
+        result = result.copy()
+
+    if convert_na_value:
+        for i in range(n):
+            val = result[i]
+            if not checknull(val):
```

Comment:
do we also need to account for pandas.options.mode.use_inf_as_na

Reply:
Not super familiar with ``pandas.options.mode.use_inf_as_na``, but it looks like it only treats inf as nan, but it does not convert, right?

Reply:
ok, support looks patchy, but does work as I expected for Int64, maybe I misunderstood the functionality, but I do expect it to convert in operations such as astype or DataFrame construction.

```
>>> import numpy as np
>>> import pandas as pd
>>>
>>> pd.__version__
'1.2.0.dev0+27.g53f6b4711'
>>>
>>> with pd.option_context("mode.use_inf_as_na", True):
...     df = pd.DataFrame({"a": np.array([np.nan, np.inf, 1.0])})
...
>>> df
     a
0  NaN
1  inf
2  1.0
>>>
>>> df = pd.DataFrame({"a": np.array([np.nan, np.inf, 1.0])})
>>> with pd.option_context("mode.use_inf_as_na", True):
...     df = df.astype("float32")
...
>>> df
     a
0  NaN
1  inf
2  1.0
>>>
>>> with pd.option_context("mode.use_inf_as_na", True):
...     df = df.astype("category")
...
Traceback (most recent call last):
...
ValueError: Categorical categories cannot be null 
>>>
>>> df = pd.DataFrame({"a": np.array([np.nan, np.inf, 1.0])})
>>> with pd.option_context("mode.use_inf_as_na", True):
...     df = df.astype("Int64")
...
>>> df
      a
0  <NA>
1  <NA>
2     1
>>>
>>> df = pd.DataFrame({"a": np.array([np.nan, np.inf, 1.0])})
>>> with pd.option_context("mode.use_inf_as_na", False):
...     df = df.astype("Int64")
...
Traceback (most recent call last):
...
TypeError: cannot safely cast non-equivalent float64 to int64
>>> df
     a
0  NaN
1  inf
2  1.0
>>>
>>> df = pd.DataFrame({"a": np.array([np.nan, np.inf, "foo"])})
>>> with pd.option_context("mode.use_inf_as_na", False):
...     df = df.astype("string")
...
>>> df
     a
0  nan
1  inf
2  foo
>>>
>>> df = pd.DataFrame({"a": np.array([np.nan, np.inf, "foo"])})
>>> with pd.option_context("mode.use_inf_as_na", True):
...     df = df.astype("string")
...
>>> df
     a
0  nan
1  inf
2  foo
>>>
>>>
```

Reply:
``pd.options.mode.use_inf_as_na = True`` doesn't change inf to nan, so I'm not sure an astype should return a string or a nan. For example:

```python
>>> pd.options.mode.use_inf_as_na = True
>>> x = pd.Series([np.inf, np.nan])
>>> x
0   NaN
1   NaN
dtype: float64
>>> x[0]
inf  # it's an inf, only the repr is a Nan it's and treated as nan in calculations...
```

Pandas 1.0 had not decided on how to convert infs to strings when ``pd.options.mode.use_inf_as_na = True``:

```python
>>> pd.options.mode.use_inf_as_na = True
>>> x = pd.Series([np.inf, np.nan])
>>> x.astype(str)
0    inf  # <- string "inf", not string "nan"
1    nan
dtype: object
>>> x.astype("string")
0    <NA>  " <- pd. NA
1    <NA>
dtype: string
```

Probably a choice of convention on how inf should be converted when ``use_inf_as_na = True`` needs to be decided on? In that case it'd be a different PR IMO.

Reply:
> Probably a choice of convention on how inf should be converted when `use_inf_as_na = True` needs to be decided on? In that case it'd be a different PR IMO.

sure